File: Ramses_vs_GOFS31.m
%% Glider track as a function of time

clear all;

%% User input

% RAMSES (MAB + SAB)
lon_lim = [-81 -70];
lat_lim = [30 42];
gdata = 'https://data.ioos.us/thredds/dodsC/deployments/secoora/ramses-20180907T0000/ramses-20180907T0000.nc3.nc';

% Folder where to save figure
folder = '/Users/aristizabal/Desktop/MARACOOS_project/Maria_scripts/Figures/Model_glider_comp/';

%%

% Ramses
url_glider = gdata;
model_name = 'GOFS 3.1';
url_model = 'http://tds.hycom.org/thredds/dodsC/GLBv0.08/expt_93.0/ts3z';
var = 'temperature';
fig = 'no';
date_ini = ' ';
date_end = '18-Sep-2018 00:00:00';

[timeg_vec, depthg_vec, tempg_matrix,timem,depthm,tempm] = ...
          glider_transect_model_comp(url_glider,model_name,url_model,var,fig,date_ini,date_end); 
      
%%      
siz_text = 20;
siz_title =20;

var_name = ncreadatt(url_glider,var,'ioos_category');
var_units = ncreadatt(url_glider,var,'units');

inst_id = ncreadatt(url_glider,'/','id');
inst = strsplit(inst_id,'-');
inst_name = inst{1};

cc_vec = floor(min(min(tempg_matrix))):1:ceil(max(max(tempg_matrix)));
    
figure
set(gcf,'position',[327 434 1301 521*2])

subplot(211)
contourf(timeg_vec,-depthg_vec,tempg_matrix,cc_vec,'.--k')
hold on
contour(timeg_vec,-depthg_vec,tempg_matrix,[26 26],'-k','linewidth',2)
shading interp

set(gca,'fontsize',siz_text)
ylabel('Depth (m)')
title(['Along track ',var_name,' Profile ',inst_name],'fontsize',siz_title)

cc = jet(length(cc_vec)-1);
colormap(cc)
c = colorbar;
c.Label.String = [var_name,' ','(',var_units,')'];
c.Label.FontSize = siz_text;
caxis([floor(min(min(tempg_matrix))) ceil(max(max(tempg_matrix)))])
set(c,'ytick',cc_vec)

tt_vec = unique(floor([timeg_vec(1),(timeg_vec(end)-timeg_vec(1))/10:(timeg_vec(end)-timeg_vec(1))/10:timeg_vec(end),timeg_vec(end)]));
xticks(tt_vec)
xticklabels(datestr(tt_vec,'mm/dd/yy'))
xlim([tt_vec(1) timeg_vec(end)])

%ylim([-max(depthg_vec) 0])
%yticks(floor(-max(depthg_vec):max(depthg_vec)/5:0))
ylim([-40 0])
yticks(-40:10:0)

%{
patch([datenum(2018,8,1,9,0,0) datenum(2018,8,1,12,0,0) datenum(2018,8,1,12,0,0) datenum(2018,8,1,9,0,0)],...
     [-300 -300 0 0],[0.5 0.5 0.5],'facealpha',0.3) 
  
patch([datenum(2018,8,2,9,0,0) datenum(2018,8,2,12,0,0) datenum(2018,8,2,12,0,0) datenum(2018,8,2,9,0,0)],...
      [-300 -300 0 0],[0.5 0.5 0.5],'facealpha',0.3) 
  
patch([datenum(2018,8,3,9,0,0) datenum(2018,8,3,12,0,0) datenum(2018,8,3,12,0,0) datenum(2018,8,3,9,0,0)],...
      [-300 -300 0 0],[0.5 0.5 0.5],'facealpha',0.3)   
%}
set(gca,'TickDir','out') 
set(gca,'xgrid','on','ygrid','on','layer','top')

ax = gca;
ax.GridAlpha = 0.3;

%%

subplot(212)
contourf(timem,-depthm,tempm,cc_vec,'.--k')
hold on
contour(timem,-depthm,tempm,[26 26],'-k','linewidth',2)
shading interp

set(gca,'fontsize',siz_text)
ylabel('Depth (m)')
title(['Along Track ',var_name,' Profile ',model_name],'fontsize',siz_title)

cc = jet(length(cc_vec)-1);
colormap(cc)
c = colorbar;
c.Label.String = [var_name,' ','(',var_units,')'];
c.Label.FontSize = siz_text;
caxis([floor(min(min(tempg_matrix))) ceil(max(max(tempg_matrix)))])
set(c,'ytick',cc_vec)

tt_vec = unique(floor([timeg_vec(1),timeg_vec(1)+(timeg_vec(end)-timeg_vec(1))/10:(timeg_vec(end)-timeg_vec(1))/10:timeg_vec(end),timeg_vec(end)]));
xticks(tt_vec)
xticklabels(datestr(tt_vec,'mm/dd/yy'))
xlim([tt_vec(1) timeg_vec(end)])

%ylim([-max(depthg_vec) 0])
%yticks(floor(-max(depthg_vec):max(depthg_vec)/5:0))
ylim([-40 0])
yticks(-40:10:0)

%{
patch([datenum(2018,8,1,9,0,0) datenum(2018,8,1,12,0,0) datenum(2018,8,1,12,0,0) datenum(2018,8,1,9,0,0)],...
     [-300 -300 0 0],[0.5 0.5 0.5],'facealpha',0.3) 
  
patch([datenum(2018,8,2,9,0,0) datenum(2018,8,2,12,0,0) datenum(2018,8,2,12,0,0) datenum(2018,8,2,9,0,0)],...
      [-300 -300 0 0],[0.5 0.5 0.5],'facealpha',0.3) 
  
patch([datenum(2018,8,3,9,0,0) datenum(2018,8,3,12,0,0) datenum(2018,8,3,12,0,0) datenum(2018,8,3,9,0,0)],...
      [-300 -300 0 0],[0.5 0.5 0.5],'facealpha',0.3)
%}
set(gca,'TickDir','out') 
set(gca,'xgrid','on','ygrid','on','layer','top')

ax = gca;
ax.GridAlpha = 0.3;      
      
% Figure name
Fig_name = [folder,'Along_track_',var,'_prof_',inst_name,'_',datestr(timem(1),'mm-dd-yy'),'-',datestr(timem(end),'mm-dd-yy')];
wysiwyg
print([Fig_name,'.png'],'-dpng','-r300') 

%%

% RU33
var = 'salinity';


[timeg_vec, depthg_vec, saltg_matrix,timem,depthm,salm] = ...
          glider_transect_model_comp(url_glider,model_name,url_model,var,fig,date_ini,date_end);
     
      
%%      
siz_text = 20;
siz_title =20;

var_name = ncreadatt(url_glider,var,'ioos_category');
var_units = ncreadatt(url_glider,var,'units');

inst_id = ncreadatt(url_glider,'/','id');
inst = strsplit(inst_id,'-');
inst_name = inst{1};

%cc_vec = floor(min(min(varg_matrix))):0.2:ceil(max(max(varg_matrix)));
cc_vec = 24:1:37;

figure
set(gcf,'position',[327 434 1301 521*2])

subplot(211)
contourf(timeg_vec,-depthg_vec,saltg_matrix)%,cc_vec)%,'.--k')
hold on
%contour(timeg_vec,-depthg_vec,saltg_matrix,[26 26],'-k','linewidth',2)
shading interp

set(gca,'fontsize',siz_text)
ylabel('Depth (m)')
title(['Along Track ',var_name,' Profile ',inst_name],'fontsize',siz_title)

cc = jet(length(cc_vec)-1);
colormap(cc)
c = colorbar;
c.Label.String = [var_name,' ','(',var_units,')'];
c.Label.FontSize = siz_text;
%caxis([floor(min(min(varg_matrix))) ceil(max(max(varg_matrix)))])
caxis([24 33])
set(c,'ytick',cc_vec)

tt_vec = unique(floor([timeg_vec(1),timeg_vec(1)+(timeg_vec(end)-timeg_vec(1))/10:(timeg_vec(end)-timeg_vec(1))/10:timeg_vec(end),timeg_vec(end)]));
xticks(tt_vec)
xticklabels(datestr(tt_vec,'mm/dd/yy'))
xlim([tt_vec(1) timeg_vec(end)])

%ylim([-max(depthg_vec) 0])
%yticks(floor(-max(depthg_vec):max(depthg_vec)/5:0))
ylim([-40 0])
yticks(-40:10:0)

%{
patch([datenum(2018,8,1,9,0,0) datenum(2018,8,1,12,0,0) datenum(2018,8,1,12,0,0) datenum(2018,8,1,9,0,0)],...
     [-300 -300 0 0],[0.5 0.5 0.5],'facealpha',0.3) 
  
patch([datenum(2018,8,2,9,0,0) datenum(2018,8,2,12,0,0) datenum(2018,8,2,12,0,0) datenum(2018,8,2,9,0,0)],...
      [-300 -300 0 0],[0.5 0.5 0.5],'facealpha',0.3) 
  
patch([datenum(2018,8,3,9,0,0) datenum(2018,8,3,12,0,0) datenum(2018,8,3,12,0,0) datenum(2018,8,3,9,0,0)],...
      [-300 -300 0 0],[0.5 0.5 0.5],'facealpha',0.3)   
%}

set(gca,'TickDir','out') 
set(gca,'xgrid','on','ygrid','on','layer','top')

ax = gca;
ax.GridAlpha = 0.3;

%%
subplot(212)
contourf(timem,-depthm,salm,cc_vec,'.--k')
hold on
contour(timem,-depthm,salm,[26 26],'-k','linewidth',2)
shading interp

set(gca,'fontsize',siz_text)
ylabel('Depth (m)')
title(['Along track ',var_name,' profile ',model_name],'fontsize',siz_title)

cc = jet(length(cc_vec)-1);
colormap(cc)
c = colorbar;
c.Label.String = [var_name,' ','(',var_units,')'];
c.Label.FontSize = siz_text;
%caxis([floor(min(min(varg_matrix))) ceil(max(max(varg_matrix)))])
caxis([24 33])
set(c,'ytick',cc_vec)

tt_vec = unique(floor([timeg_vec(1),timeg_vec(1)+(timeg_vec(end)-timeg_vec(1))/10:(timeg_vec(end)-timeg_vec(1))/10:timeg_vec(end),timeg_vec(end)]));
xticks(tt_vec)
xticklabels(datestr(tt_vec,'mm/dd/yy'))
xlim([tt_vec(1) timeg_vec(end)])

%ylim([-max(depthg_vec) 0])
%yticks(floor(-max(depthg_vec):max(depthg_vec)/5:0))
ylim([-40 0])
yticks(-40:10:0)

%{
patch([datenum(2018,8,1,9,0,0) datenum(2018,8,1,12,0,0) datenum(2018,8,1,12,0,0) datenum(2018,8,1,9,0,0)],...
     [-300 -300 0 0],[0.5 0.5 0.5],'facealpha',0.3) 
  
patch([datenum(2018,8,2,9,0,0) datenum(2018,8,2,12,0,0) datenum(2018,8,2,12,0,0) datenum(2018,8,2,9,0,0)],...
      [-300 -300 0 0],[0.5 0.5 0.5],'facealpha',0.3) 
  
patch([datenum(2018,8,3,9,0,0) datenum(2018,8,3,12,0,0) datenum(2018,8,3,12,0,0) datenum(2018,8,3,9,0,0)],...
      [-300 -300 0 0],[0.5 0.5 0.5],'facealpha',0.3)   
%}

set(gca,'TickDir','out') 
set(gca,'xgrid','on','ygrid','on','layer','top')

ax = gca;
ax.GridAlpha = 0.3;      
      
% Figure name
Fig_name = [folder,'Along_track_',var,'_prof_',inst_name,'_',datestr(timem(1),'mm-dd-yy'),'-',datestr(timem(end),'mm-dd-yy')];
wysiwyg
print([Fig_name,'.png'],'-dpng','-r300') 

%%
save('Ramses_GOFS31.mat')

File: Ramses_vs_GOFS31.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed May  8 15:15:25 2019

@author: aristizabal
"""

#%% User input

# ramses
gdata = 'https://gliders.ioos.us/thredds/dodsC/deployments/secoora/ramses-20180907T0000/ramses-20180907T0000.nc3.nc'

# date limits
date_ini = '2018-09-08T00:00:00Z'
date_end = '2018-09-18T00:00:00Z'

mat_file = '/Users/aristizabal/Desktop/MARACOOS_project/Maria_scripts/Figures/Model_glider_comp/Ramses_GOFS31.mat'

# Folder where to save figure
folder = '/Users/aristizabal/Desktop/MARACOOS_project/Maria_scripts/Figures/Model_glider_comp/'

#%%

import numpy as np
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
from datetime import datetime, timedelta  
import xarray as xr
import netCDF4
from matplotlib.dates import date2num 
import cmocean

import scipy.io as sio
ramses = sio.loadmat(mat_file)

ramses.keys()

#%% Downloding data

depthg_vec_ram = ramses['depthg_vec'][0,:]
tempg_matrix_ram = ramses['tempg_matrix'][:]
tstamp_glider_ram = ramses['timeg_vec'][:,0]

depth31 = ramses['depthm'][:,0]
temp31 = ramses['tempm'][:]
salt31 = ramses['salm'][:]
tstamp31 = ramses['timem'][:,0]

#%% Changing timestamps to datenum

timeglid = []
for i in np.arange(len(tstamp_glider_ram)):
    timeglid.append(datetime.fromordinal(int(tstamp_glider_ram[i])) + \
        timedelta(days=tstamp_glider_ram[i]%1) - timedelta(days = 366))
timeglider_ram = np.asarray(timeglid)
    
timeglid = []
for i in np.arange(len(tstamp31)):
    timeglid.append(datetime.fromordinal(int(tstamp31[i])) + \
        timedelta(days=tstamp31[i]%1) - timedelta(days = 366))
time31 = np.asarray(timeglid)

#%% Reading glider data

ncglider = xr.open_dataset(gdata,decode_times=False)
latglider = ncglider.latitude[:]
longlider = ncglider.longitude[:]
time_glider = ncglider.time
time_glider = netCDF4.num2date(time_glider[:],time_glider.units)
tempglider = np.array(ncglider.temperature[0,:,:])
saltglider = np.array(ncglider.salinity[0,:,:])
depthglider = np.array(ncglider.depth[0,:,:])

timestamp_glider = date2num(time_glider)[0]

#%%
tmin = datetime.strptime(date_ini,'%Y-%m-%dT%H:%M:%SZ')
tmax = datetime.strptime(date_end,'%Y-%m-%dT%H:%M:%SZ')

okg = np.where(np.logical_and(time_glider.T >= tmin, time_glider.T <= tmax))

timeg = time_glider[0,okg[0]]
timestampg = timestamp_glider[okg[0]]
latg = np.asarray(latglider[0,okg[0]])
long = np.asarray(longlider[0,okg[0]])
depthg = depthglider[okg[0],:]
tempg = tempglider[okg[0],:]
saltg = saltglider[okg[0],:]
          
#%% ramses contour plot

timegg = timeglider_ram
depthg_gridded = depthg_vec_ram
varg_gridded = tempg_matrix_ram
inst_id = 'Ramses'

nlevels =21
kw = dict(levels = np.linspace(9,29,nlevels))

fig, ax=plt.subplots(figsize=(12, 3))
plt.contour(timegg,-depthg_gridded.T,varg_gridded,[26],colors = 'k')
cs = plt.contourf(timegg,-depthg_gridded.T,varg_gridded,cmap='RdYlBu_r',**kw)
plt.title(inst_id.split('-')[0],fontsize=20)

ax.set_xlim(timegg[0], timegg[-1])
xfmt = mdates.DateFormatter('%d-%b\n %Y')
ax.xaxis.set_major_formatter(xfmt)

cbar = fig.colorbar(cs, orientation='vertical') 
cbar.ax.set_ylabel('Temperature ($^oC$)',fontsize=16)
cbar.ax.tick_params(labelsize=14)
ax.set_ylabel('Depth (m)',fontsize=16);
plt.xticks(fontsize=16)
plt.yticks(fontsize=14) 
#plt.xlim([datetime(2018,7,18),datetime(2018,9,16)])
plt.ylim([-40,0]) 
plt.xticks([datetime(2018,9,8),datetime(2018,9,10),datetime(2018,9,12),\
            datetime(2018,9,14),datetime(2018,9,16)])
plt.yticks([-40,-30,-20,-10,0])                                      

file = "/Users/aristizabal/Desktop/MARACOOS_project/Maria_scripts/Figures/Model_glider_comp/ramses_temp.png"
plt.savefig(file,bbox_inches = 'tight',pad_inches = 0.1) 

#%% GOFS 3.1 temperature

nlevels =17
kw = dict(levels = np.linspace(13,29,nlevels))
    
fig, ax=plt.subplots(figsize=(10, 1.7))
plt.contour(time31,-1*depth31.T,temp31,[26],colors = 'k')
cs = plt.contourf(time31,-1*depth31.T,temp31,cmap='RdYlBu_r',**kw)
plt.plot(np.tile(datetime(2018,9,14,00),len(np.arange(-40,0))),np.arange(-40,0),'--k')
plt.title('GOFS 3.1',fontsize=20)
ax.set_xlim(timeg[0], timeg[-1])
xfmt = mdates.DateFormatter('%d-%b\n %Y')
ax.xaxis.set_major_formatter(xfmt)

cbar = fig.colorbar(cs, orientation='vertical') 
cbar.ax.set_ylabel('Temperature ($^oC$)',fontsize=16)
cbar.ax.tick_params(labelsize=14)
ax.set_ylabel('Depth (m)',fontsize=16);
plt.xticks(fontsize=16)
plt.yticks(fontsize=14) 
xfmt = mdates.DateFormatter('%H:%Mh\n%d-%b')
ax.xaxis.set_major_formatter(xfmt)
plt.ylim([-40,0]) 
plt.yticks([-40,-20,0]) 
plt.xticks([datetime(2018,9,8),datetime(2018,9,10),datetime(2018,9,12),\
            datetime(2018,9,14),datetime(2018,9,16),datetime(2018,9,18)])
#plt.yticks([-40,-30,-20,-10,0])                                      

file = "/Users/aristizabal/Desktop/MARACOOS_project/Maria_scripts/Figures/Model_glider_comp/ramses_GOFS31_temp.png"
plt.savefig(file,bbox_inches = 'tight',pad_inches = 0.1) 

#%% GOFS 3.1 salnity

nlevels =11
kw = dict(levels = np.linspace(26,36,nlevels))
    
fig, ax=plt.subplots(figsize=(10, 1.7))
cs = plt.contourf(time31,-1*depth31.T,salt31,cmap=cmocean.cm.haline,**kw)
plt.plot(np.tile(datetime(2018,9,14,00),len(np.arange(-40,0))),np.arange(-40,0),'--k')
plt.title('GOFS 3.1',fontsize=20)
ax.set_xlim(timeg[0], timeg[-1])
xfmt = mdates.DateFormatter('%d-%b\n %Y')
ax.xaxis.set_major_formatter(xfmt)

cbar = fig.colorbar(cs, orientation='vertical') 
cbar.ax.set_ylabel('Salinity',fontsize=16)
cbar.ax.tick_params(labelsize=14)
ax.set_ylabel('Depth (m)',fontsize=16);
plt.xticks(fontsize=16)
plt.yticks(fontsize=14) 
xfmt = mdates.DateFormatter('%H:%Mh\n%d-%b')
ax.xaxis.set_major_formatter(xfmt)
plt.ylim([-40,0]) 
plt.yticks([-40,-20,0]) 
plt.xticks([datetime(2018,9,8),datetime(2018,9,10),datetime(2018,9,12),\
            datetime(2018,9,14),datetime(2018,9,16),datetime(2018,9,18)])
                           

file = "/Users/aristizabal/Desktop/MARACOOS_project/Maria_scripts/Figures/Model_glider_comp/ramses_GOFS31_salt.png"
plt.savefig(file,bbox_inches = 'tight',pad_inches = 0.1) 


#%% ramses scatter temperature

timeg_matrix = np.transpose(np.tile(timeg.T,(depthg.shape[1],1)))
ttg = np.ravel(timeg_matrix)
dg = np.ravel(depthg)
teg = np.ravel(tempg)

kw = dict(c=teg, marker='*', edgecolor='none')

fig, ax = plt.subplots(figsize=(10, 1.7))
cs = ax.scatter(ttg,-dg,cmap='RdYlBu_r',**kw)
plt.contour(timegg,-depthg_gridded.T,varg_gridded,[26],colors = 'k')
plt.plot(np.tile(datetime(2018,9,14,00),len(np.arange(-40,0))),np.arange(-40,0),'--k')
ax.set_xlim(timeg[0], timeg[-1])
ax.set_ylabel('Depth (m)',fontsize=16)
cbar = plt.colorbar(cs)
cbar.ax.set_ylabel('Temperature ($^oC$)',fontsize=16)
cbar.set_ticks(np.arange(13,31,2))
cbar.ax.tick_params(labelsize=14)

cbar.set_clim(13,29)
ax.set_title(inst_id.split('-')[0],fontsize=20)
xfmt = mdates.DateFormatter('%H:%Mh\n%d-%b')
ax.xaxis.set_major_formatter(xfmt)
plt.xticks(fontsize=16)
plt.yticks(fontsize=14) 
plt.xticks([datetime(2018,9,8),datetime(2018,9,10),datetime(2018,9,12),\
            datetime(2018,9,14),datetime(2018,9,16),datetime(2018,9,18)])
    
file = "/Users/aristizabal/Desktop/MARACOOS_project/Maria_scripts/Figures/Model_glider_comp/ramses_temp_scatter.png"
plt.savefig(file,bbox_inches = 'tight',pad_inches = 0.1) 

#%% ramses scatter salinity

timeg_matrix = np.transpose(np.tile(timeg.T,(depthg.shape[1],1)))
ttg = np.ravel(timeg_matrix)
dg = np.ravel(depthg)
seg = np.ravel(saltg)

kw = dict(c=seg, marker='*', edgecolor='none')

fig, ax = plt.subplots(figsize=(10, 1.7))
cs = ax.scatter(ttg,-dg,cmap=cmocean.cm.haline,**kw)
plt.plot(np.tile(datetime(2018,9,13,18),len(np.arange(-40,0))),np.arange(-40,0),'--k')
ax.set_xlim(timeg[0], timeg[-1])
ax.set_ylabel('Depth (m)',fontsize=16)

cbar = plt.colorbar(cs)
cbar.ax.set_ylabel('Salinity',fontsize=16)
cbar.set_ticks(np.arange(26,38,2))
cbar.ax.tick_params(labelsize=14)
cbar.set_clim(26,36)

ax.set_title(inst_id.split('-')[0],fontsize=20)
xfmt = mdates.DateFormatter('%H:%Mh\n%d-%b')
ax.xaxis.set_major_formatter(xfmt)
plt.xticks(fontsize=16)
plt.yticks(fontsize=14) 
plt.xticks([datetime(2018,9,8),datetime(2018,9,10),datetime(2018,9,12),\
            datetime(2018,9,14),datetime(2018,9,16),datetime(2018,9,18)])
    
file = "/Users/aristizabal/Desktop/MARACOOS_project/Maria_scripts/Figures/Model_glider_comp/ramses_salt_scatter.png"
plt.savefig(file,bbox_inches = 'tight',pad_inches = 0.1)     